### PR TITLE
 LibPDF: Add basic tiled, coloured pattern rendering 

### PR DESCRIPTION
**Pull request by X-yl**
PDF.js vs LibPDF :)

Tests/LibPDF/pattern.pdf_p1_ref_pdfjs
Tests/LibPDF/pattern.pdf_p1_step1

So I've added some basic pattern rendering to LibPDF, mainly targeting Adobe's example on p176 of the spec (pictured above). There's a couple structural-y changes that I'm not too sure about:

  - Changing all the ColorSpaces to output PaintStyles instead of Colors: This is so the pattern space can output a PaintStyle that generates the bitmap, but it is a bit overkill for the normal device spaces that have to output a SolidColorPaintStyle.
  - The PatternColorSpace sort of blends in with the Renderer code a bit since it has to render the pattern, so I decided to make it a friend class and pass in a reference to Renderer. Not too sure how to structure it tbh.

Also my C++ is a bit rusty (in more ways than one) so let me know if there's better ways to do things :)

### CODE REVIEW

**Comment by X-yl** [comment]
Hmm discovered by tweaking the text matrix of he file I can stop the text clipping: 

Tests/LibPDF/pattern.pdf_p1_after

With a change like this:
```diff
--- a/Tests/LibPDF/pattern.pdf
+++ b/Tests/LibPDF/pattern.pdf
@@ -23,7 +23,7 @@ stream
 BT % Begin text object
 0.0 -0.05 TD % Move text position
 /F1 1 Tf % Set text font and size
-55 0 0 55 7.1771 2.4414 Tm % Set text matrix
+55 0 0 55 7.1771 10.4414 Tm % Set text matrix
 0 Tc % Set character spacing
 0 Tw % Set word spacing
 1.0 0.0 0.0 rg % Set nonstroking colour to red
```

But I'm wondering if that's a bug or if we expect the text matrix to be different because of different fonts or something?

**Review by nico — COMMENTED** [review]
This is is super cool!

The PaintStyle change looks extremely clean. However, we use the ColorSpace machinery for transforming images too (cf `Renderer::load_image()` towards the end) and I worry a bit that creating one PaintStyle object _per call to `style()`_ will make performance even harder to optimize.

Maybe we should instead just special-case Pattern in Renderer more and do a bunch of `is pattern then call style else call color` or something like that?

Also, do you think we could pass Renderer to `create()` and have the Pattern object remember it, instead of passing it to `style()` all the time?

Finally, the "preferences" object is for user-settable things I think. Maybe we can add a parameter for the bg color?

**Comment by X-yl** [comment]
Perhaps we could have the style function return a `Variant<Color, RefPtr<PaintStyle>>` (typedef'd to something less clunky ideally) and then branch on whether it produced a `RefPtr<PaintStyle>` or a `Color`?

**Comment by X-yl** [comment]
Made changes based on feedback :)

I do recognise the difference in semantics between checking if the `ColorSpace` is pattern and then calling `style()` or else `color()` versus having `style()` returning `Variant<Color, PaintStyle>` but I think it's a bit cleaner (as opposed to stubbing out `style()` for almost every `ColorSpace`) and performance wise is they should be exactly the same since the branch predictor would quite easily predict what each `ColorSpace` returns?

File: Userland/Libraries/LibPDF/ColorSpace.cpp [hunk 1]
```diff
@@ -101,7 +101,7 @@ NonnullRefPtr<DeviceGrayColorSpace> DeviceGrayColorSpace::the()
     return instance;
 }
 
-PDFErrorOr<Color> DeviceGrayColorSpace::color(ReadonlySpan<Value> arguments) const
+PDFErrorOr<ColorOrStyle> DeviceGrayColorSpace::style(ReadonlySpan<Value> arguments) const
```
Comment (nico) [inline Userland/Libraries/LibPDF/ColorSpace.cpp:102]: Another idea could be to make this return a reference to a Style and make the non-Pattern colorspaces could always return the same object and just change the value it returns?

Or we could just special-case Pattern. The callsites are already a mess anyways.

Or we could get back to your original way and worry about perf later.

But as-is is fine too, whatever you think is best.

File: Userland/Libraries/LibPDF/Renderer.cpp [hunk 2]
```diff
@@ -293,6 +293,11 @@ RENDERER_HANDLER(path_append_rect)
 
 void Renderer::begin_path_paint()
 {
+    if (state().paint_style.has<NonnullRefPtr<Gfx::PaintStyle>>()) {
+        m_original_paint_style = state().paint_style.get<NonnullRefPtr<Gfx::PaintStyle>>();
```
Comment (nico) [inline Userland/Libraries/LibPDF/Renderer.cpp:297]: Add a `VERIFY(!m_original_paint_style)` before this line.

**Review by nico — APPROVED** [review]
This is great!

lgtm as-is; some pontificating below (but feel free to ignore; up to you).

**Comment by nico** [comment]
Also, I don't remember if Serenity's PDF viewer supports page zoom (MacPDF doesn't yet). Might be worth thinking about if the pattern rasterization applies zoom at the right layer, so that zooming in doesn't make the pattern pixelated.

File: Userland/Libraries/LibPDF/ColorSpace.cpp (hunk 1 above)
Comment (X-yl) [reply Userland/Libraries/LibPDF/ColorSpace.cpp:102]: I wasn't too happy with using `SolidColorPaintStyle` everywhere either, it seemed a bit overkill. I think `Variant<Color, PaintStyle>` is probably a natural way to go, doesn't immediately mean perf will be terrible and also doesn't blow up the inheritance hierarchy (which is possibly a bad thing - but at least it mirrors the spec). I think just roll with it for now though, we can always come back and refactor.

**Comment by X-yl** [comment]
> Also, I don't remember if Serenity's PDF viewer supports page zoom (MacPDF doesn't yet). Might be worth thinking about if the pattern rasterization applies zoom at the right layer, so that zooming in doesn't make the pattern pixelated.

So the way I decided on the size for the bitmap for rasterization was to take the pattern bounding box and transform it to device space with the CTM (ETA: after first mapping it to user space with the pattern matrix). This _seems_ to handle scaling nicely, and intuitively I think it's right, but frankly I did a lot of throwing random matrices at the wall to see what stuck so I could be missing something. 

Tests/LibPDF/pattern.pdf_p1_ref_zoomed in

**Review by nico — APPROVED** [review]

**Pull request merged**